src/System.Management.Automation/engine/Modules/ModuleUtils.cs

```diff
@@ -106,6 +111,110 @@ internal static IEnumerable<string> GetAllAvailableModuleFiles(string topDirecto
+            }
+
+            // Work out the psd1 path
+            string dirName = Path.GetFileName(directoryPath.TrimEnd('/', '\\'));
+            string manifestPath = Path.Join(directoryPath, dirName + ".psd1");
+
+            // No manifest means this is not a module directory
+            if (!File.Exists(manifestPath))
+            {
+                return false;
+            }
+
+            // Check the manifest file for compatibility
+            return !Utils.IsPSEditionSupported(ReadCompatiblePSEditionsFromManifest(manifestPath));
+        }
+
+        /// <summary>
+        /// Try to read in the CompatiblePSEditions from a module file, lazily.
+        /// </summary>
+        /// <param name="manifestPath">The path to the module manifest to proces.</param>
+        /// <returns>All PSEditions listed as compatible by the manifest, if any.</returns>
+        private static IEnumerable<string> ReadCompatiblePSEditionsFromManifest(string manifestPath)
+        {
+            Hashtable manifest;
+            try
+            {
+                manifest = PsUtils.GetModuleManifestProperties(manifestPath, new [] { "CompatiblePSEditions" });
+            }
+            catch (Exception e) when (e is IOException || e is UnauthorizedAccessException || e is FileNotFoundException)
+            {
+                // If the file cannot be accessed, treat it as if it doesn't exist
+                yield break;
+            }
+
+            object psEditionsObj = manifest?["CompatiblePSEditions"];
+            if (psEditionsObj == null)
+            {
+                // No compatibility field => no supported editions
+                yield break;
+            }
+
+            string[] psEditions;
+            try
+            {
+                psEditions = LanguagePrimitives.ConvertTo<string[]>(psEditionsObj);
+            }
+            catch (PSInvalidCastException)
+            {
+                // If the field exists but can't be cast to a string array, ignore it
+                yield break;
+            }
+
+            if (psEditions == null)
+            {
+                // The field exists but is null => ignore it
+                yield break;
+            }
+
+            // Finally return the supported editions entry
+            foreach (string psEdition in psEditions)
+            {
+                yield return psEdition;
+            }
+        }
+
+        /// <summary>
+        /// Check if the CompatiblePSEditions field of a given module
+        /// declares compatibility with the running PowerShell edition.
+        /// </summary>
+        /// <param name="moduleManifestPath">The path to the module manifest being checked.</param>
+        /// <param name="compatiblePSEditions">The value of the CompatiblePSEditions field of the module manifest.</param>
+        /// <returns>True if the module is compatible with the running PowerShell edition, false otherwise.</returns>
+        internal static bool IsPSEditionCompatible(
```

Comment:
The name should be `IsModuleConsideredCompatible`, because a module doesn't necessarily need to declare `CompatiblePSEditions` for it to be considered compatible.